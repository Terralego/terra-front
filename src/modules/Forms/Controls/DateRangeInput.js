import React from 'react';
import classnames from 'classnames';

import PropTypes from 'prop-types';

import { Spinner } from '@blueprintjs/core';
import { DateRangeInput as BPDateRangeInput } from '@blueprintjs/datetime';

import localeUtils from '../../../utils/localeUtils';
import translateMock from '../../../utils/translate';

// Date Format
const formatDate = date => date.toLocaleDateString();
const parseDate = str => new Date(str);

export const DateRangeInput = ({
  className,
  label,
  value,
  onChange,
  translate,
  locale = global.navigator.language,
  shortcuts = false,
  loading,
  ...props
}) => (
  <div
    className="control-container"
  >
    <p className="control-label">{label}</p>
    {!!loading && <Spinner size={20} />}
    {!loading && (
      <BPDateRangeInput
        className={classnames('tf-control-date-range', className)}
        formatDate={formatDate}
        onChange={onChange}
        parseDate={parseDate}
        value={value}
        overlappingDatesMessage={translate('terralego.forms.controls.daterangeinput.overlapping_dates')}
        invalidDateMessage={translate('terralego.forms.controls.daterangeinput.invalid_date')}
        popoverProps={{ usePortal: false }}
        localeUtils={localeUtils}
        locale={locale.substr(0, 2)}
        shortcuts={shortcuts}
        startInputProps={{
          placeholder: translate('terralego.forms.controls.daterangeinput.start_placeholder'),
          className: 'tf-control-date-range__start-input',
        }}
        endInputProps={{
          placeholder: translate('terralego.forms.controls.daterangeinput.end_placeholder'),
          className: 'tf-control-date-range__en d-input',
        }}
        {...props}
      />
    )}
  </div>
);

DateRangeInput.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.arrayOf(PropTypes.instanceOf(Date)),
  translate: PropTypes.func,
  loading: PropTypes.bool,
};

DateRangeInput.defaultProps = {
  label: '',
  value: [null, null],
  onChange () {},
  translate: translateMock({
    'terralego.forms.controls.daterangeinput.overlapping_dates': 'Overlapping date',
    'terralego.forms.controls.daterangeinput.invalid_date': 'Invalid date',
    'terralego.forms.controls.daterangeinput.start_placeholder': 'Start date',
    'terralego.forms.controls.daterangeinput.end_placeholder': 'End date',
  }),
  loading: false,
};

export default DateRangeInput;
