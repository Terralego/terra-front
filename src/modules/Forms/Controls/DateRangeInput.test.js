import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { DateRangeInput } from './DateRangeInput';

jest.mock('@blueprintjs/datetime', () => ({
  DateRangeInput: () => <p>BPDateRangeInput</p>,
  Spinner: () => <p>Spinner</p>,
}));

it('should render', () => {
  const tree = renderer.create((
    <DateRangeInput
      label="My range"
      onChange={() => null}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render loading', () => {
  const tree = renderer.create((
    <DateRangeInput label="Pwout" loading />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render local error', () => {
  const wrapper = shallow((
    <DateRangeInput
      label="Change Values of my range"
      onChange={() => null}
    />
  ));
  expect(wrapper.find('DateRangeInput').props().overlappingDatesMessage).toEqual('Overlapping date');
  expect(wrapper.find('DateRangeInput').props().invalidDateMessage).toEqual('Invalid date');
});

it('should mount & update correctly', () => {
  const onChange = jest.fn();
  const date = new Date();
  const wrapper = shallow((
    <DateRangeInput
      label="Change Values of my range"
      onChange={onChange}
    />
  ));
  wrapper.find('DateRangeInput').props().onChange([date, date]);
  expect(onChange).toHaveBeenCalled();
});

it('shoud render formatDate', () => {
  const date = new Date();
  const wrapper = shallow(<DateRangeInput />);
  const { formatDate } = wrapper.find('DateRangeInput').props();
  wrapper.find('DateRangeInput').props();
  expect(formatDate(date)).toBe(date.toLocaleDateString());
});

it('shoud render date', () => {
  const str = (new Date()).toLocaleDateString();
  const wrapper = shallow(<DateRangeInput />);
  const { parseDate } = wrapper.find('DateRangeInput').props();
  wrapper.find('DateRangeInput').props();
  expect(parseDate(str)).toEqual(new Date(str));
});
