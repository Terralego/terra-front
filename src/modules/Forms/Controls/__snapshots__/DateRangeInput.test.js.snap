// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="control-container"
>
  <p
    className="control-label"
  >
    My range
  </p>
  <p>
    BPDateRangeInput
  </p>
</div>
`;

exports[`should render loading 1`] = `
<div
  className="control-container"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <div
    className="bp3-spinner"
  >
    <div
      className="bp3-spinner-animation"
    >
      <svg
        height={20}
        strokeWidth="16.00"
        viewBox="-3.00 -3.00 106.00 106.00"
        width={20}
      >
        <path
          className="bp3-spinner-track"
          d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
        />
        <path
          className="bp3-spinner-head"
          d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
          pathLength={280}
          strokeDasharray="280 280"
          strokeDashoffset={210}
        />
      </svg>
    </div>
  </div>
</div>
`;
